File: Neocom/NCDatabaseCertificatesViewController.m
//
//  NCDatabaseCertificatesViewController.m
//  Neocom
//
//  Created by Артем Шиманский on 22.01.14.
//  Copyright (c) 2014 Artem Shimanski. All rights reserved.
//

#import "NCDatabaseCertificatesViewController.h"
#import "NCTableViewCell.h"
#import <objc/runtime.h>
#import "NSString+Neocom.h"
#import "NCDatabaseCertificateInfoViewController.h"

@interface NCDatabaseCertificatesViewControllerRow : NSObject
@property (nonatomic, strong) NSString* title;
@property (nonatomic, strong) NSString* subtitle;
@property (nonatomic, strong) NSManagedObjectID* iconID;
@property (nonatomic, strong) NCDBEveIcon* icon;
@property (nonatomic, strong) id object;
@end

@implementation NCDatabaseCertificatesViewControllerRow

@end

@interface NCDatabaseCertificatesViewController ()
@property (nonatomic, strong) NSArray* rows;

- (void) reload;

@end

@implementation NCDatabaseCertificatesViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
	if (self.groupID) {
		NCDBInvGroup* group = (NCDBInvGroup*) [self.databaseManagedObjectContext objectWithID:self.groupID];
		self.title = group.groupName;
	}
	self.refreshControl = nil;
	
	[self reload];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

- (void) prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
	if ([segue.identifier isEqualToString:@"NCDatabaseCertificatesViewController"]) {
		id row = [sender object];
		
		NCDatabaseCertificatesViewController* destinationViewController = segue.destinationViewController;
		destinationViewController.groupID = row;
	}
	else {
		NCDatabaseCertificateInfoViewController* controller;
		if ([segue.destinationViewController isKindOfClass:[UINavigationController class]])
			controller = [segue.destinationViewController viewControllers][0];
		else
			controller = segue.destinationViewController;

		id row = [sender object];
		controller.certificateID = row;
	}
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.rows.count;
}


#pragma mark - NCTableViewController

- (void) didChangeAccount:(NSNotification *)notification {
	[super didChangeAccount:notification];
	if ([self isViewLoaded])
		[self reload];
}

- (void) didChangeStorage:(NSNotification *)notification {
	[super didChangeStorage:notification];
	[self reload];
}

- (NSString*)tableView:(UITableView *)tableView cellIdentifierForRowAtIndexPath:(NSIndexPath *)indexPath {
	NCDatabaseCertificatesViewControllerRow* row = self.rows[indexPath.row];
	if (NSClassFromString([[row.object entity] managedObjectClassName]) == [NCDBCertCertificate class])
		return @"CertificateCell";
	else
		return @"GroupCell";
}

- (void) tableView:(UITableView *)tableView configureCell:(UITableViewCell*) tableViewCell forRowAtIndexPath:(NSIndexPath*) indexPath {
	NCDatabaseCertificatesViewControllerRow* row = self.rows[indexPath.row];
	NCDefaultTableViewCell *cell = (NCDefaultTableViewCell*) tableViewCell;
	
	if (!row.icon && row.iconID)
		row.icon = (NCDBEveIcon*) [self.databaseManagedObjectContext objectWithID:row.iconID];
	
	cell.titleLabel.text = row.title;
	cell.subtitleLabel.text = row.subtitle;
	cell.iconView.image = row.icon.image.image;
	cell.object = row.object;
}

#pragma mark - Private

- (void) reload {
	NCAccount *account = [NCAccount currentAccount];
	
	void (^load)(EVECharacterSheet*) = ^(EVECharacterSheet* characterSheet) {
		NSManagedObjectContext* managedObjectContext = [[NCDatabase sharedDatabase] createManagedObjectContext];
		[managedObjectContext performBlock:^{
			NSMutableArray* rows = [NSMutableArray new];
			
			if (self.groupID) {
				NCDBInvGroup* group = (NCDBInvGroup*) [managedObjectContext objectWithID:self.groupID];
				NSArray* certificates = [group.certificates sortedArrayUsingDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"certificateName" ascending:YES]]];
				NCDBEveIcon* unclaimedIcon = [managedObjectContext certificateUnclaimedIcon];
				for (NCDBCertCertificate* certificate in certificates) {
					
					NCTrainingQueue* trainingQueue = nil;
					NCDBCertMasteryLevel* level;
					for (NCDBCertMastery* mastery in [certificate.masteries sortedArrayUsingDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"level.level" ascending:YES]]]) {
						trainingQueue = [[NCTrainingQueue alloc] initWithCharacterSheet:characterSheet databaseManagedObjectContext:managedObjectContext];
						[trainingQueue addMastery:mastery];
						if (trainingQueue.trainingTime > 0.0)
							break;
						level = mastery.level;
					}
					NSTimeInterval trainingTime = trainingQueue.trainingTime;
					
					NCDatabaseCertificatesViewControllerRow* row = [NCDatabaseCertificatesViewControllerRow new];
					row.title = certificate.certificateName;
					if (trainingTime > 0)
						row.subtitle = [NSString stringWithFormat:NSLocalizedString(@"%@ to level %d", nil), [NSString stringWithTimeLeft:trainingTime], level ? level.level + 2 : 1];
					row.iconID = level ? [level.icon objectID] : [unclaimedIcon objectID];
					row.object = [certificate objectID];
					[rows addObject:row];
				}
			}
			else {
				NSFetchRequest* request = [NSFetchRequest fetchRequestWithEntityName:@"InvGroup"];
				request.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"groupName" ascending:YES]];
				request.predicate = [NSPredicate predicateWithFormat:@"certificates.@count > 0"];
				NSArray* groups = [managedObjectContext executeFetchRequest:request error:nil];
				NCDBEveIcon* defaultGroupIcon = [managedObjectContext defaultGroupIcon];

				for (NCDBInvGroup* group in groups) {
					NCDatabaseCertificatesViewControllerRow* row = [NCDatabaseCertificatesViewControllerRow new];
					row.title = group.groupName;
					row.iconID = group.icon ? [group.icon objectID] : [defaultGroupIcon objectID];
					row.object = [group objectID];
					[rows addObject:row];
				}
			}
			
			dispatch_async(dispatch_get_main_queue(), ^{
				self.rows = rows;
				[self.tableView reloadData];
			});
		}];
	};
	
	if (account) {
		[account loadCharacterSheetWithCompletionBlock:^(EVECharacterSheet *characterSheet, NSError *error) {
			load(characterSheet);
		}];
	}
	else
		load(nil);
}

@end

File: Legacy/Neocom/Neocom/NCDatabaseCertificatesViewController.swift
//
//  NCDatabaseCertificatesViewController.swift
//  Neocom
//
//  Created by Artem Shimanski on 26.12.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

import UIKit
import CoreData
import EVEAPI
import Futures

fileprivate class NCDatabaseCertificateRow: NCFetchedResultsObjectNode<NCDBCertCertificate> {
	var character: NCCharacter?
	var image: UIImage?
	var subtitle: String?
	
	required init(object: NCDBCertCertificate) {
		super.init(object: object)
		cellIdentifier = Prototype.NCDefaultTableViewCell.default.reuseIdentifier
	}
	
	override func configure(cell: UITableViewCell) {
		guard let cell = cell as? NCDefaultTableViewCell else {return}
		
		cell.object = object
		cell.titleLabel?.text = object.certificateName
		cell.accessoryType = .disclosureIndicator

		if let subtitle = subtitle {
			cell.iconView?.image = image
			cell.subtitleLabel?.text = subtitle
		}
		else  {
			cell.iconView?.image = NCDatabase.sharedDatabase?.eveIcons[NCDBEveIcon.File.certificateUnclaimed.rawValue]?.image?.image
			cell.subtitleLabel?.text = nil
			
			if let character = character {
				NCDatabase.sharedDatabase?.performBackgroundTask{ managedObjectContext in
					let certificate = (try! managedObjectContext.existingObject(with: self.object.objectID)) as! NCDBCertCertificate
					let trainingQueue = NCTrainingQueue(character: character)
					var level: NCDBCertMasteryLevel?
					for mastery in (certificate.masteries?.sortedArray(using: [NSSortDescriptor(key: "level.level", ascending: true)]) as? [NCDBCertMastery]) ?? [] {
						trainingQueue.add(mastery: mastery)
						if !trainingQueue.skills.isEmpty {
							break
						}
						level = mastery.level
					}
					let trainingTime = trainingQueue.trainingTime(characterAttributes: character.attributes)
					let subtitle: String
					let image: UIImage?
					
					if trainingTime > 0 {
						subtitle = String(format: NSLocalizedString("%@ to level %d", comment: ""), NCTimeIntervalFormatter.localizedString(from: trainingTime, precision: .seconds), (level?.level ?? -1) + 2)
					}
					else {
						subtitle = ""
					}
					image = level?.icon?.image?.image ?? NCDBEveIcon.eveIcons(managedObjectContext: managedObjectContext)[NCDBEveIcon.File.certificateUnclaimed.rawValue]?.image?.image
					
					DispatchQueue.main.async {
						self.image = image
						self.subtitle = subtitle
						self.treeController?.reloadCells(for: [self])

//						if (cell.object as? NCDBCertCertificate) == self.object {
//							cell.iconView?.image = image
//							cell.subtitleLabel?.text = subtitle
//						}
					}
				}
			}
		}
	}
}

class NCDatabaseCertificatesViewController: NCTreeViewController {

	var group: NCDBInvGroup?
	
	override func viewDidLoad() {
		super.viewDidLoad()
		tableView.register([Prototype.NCDefaultTableViewCell.default])
		title = group?.groupName
	}
	
	override func didReceiveMemoryWarning() {
		if !isViewLoaded || view.window == nil {
			treeController?.content = nil
		}
	}
	
	override func content() -> Future<TreeNode?> {
		guard let group = group else {return .init(nil)}
		
		let request = NSFetchRequest<NCDBCertCertificate>(entityName: "CertCertificate")
		request.predicate = NSPredicate(format: "group == %@", group)
		request.sortDescriptors = [NSSortDescriptor(key: "certificateName", ascending: true)]
		
		let results = NSFetchedResultsController(fetchRequest: request, managedObjectContext: NCDatabase.sharedDatabase!.viewContext, sectionNameKeyPath: nil, cacheName: nil)


		Progress(totalUnitCount: 1).perform {
			NCCharacter.load(account: NCAccount.current).then(on: .main) { character in
				self.treeController?.content?.children.forEach {
					($0 as? NCDatabaseCertificateRow)?.character = character
				}
			}.catch(on: .main) { _ in
				let character = NCCharacter()
				self.treeController?.content?.children.forEach {
					($0 as? NCDatabaseCertificateRow)?.character = character
				}
			}.finally(on: .main) {
				self.tableView.reloadData()
			}
		}
		
		return .init(FetchedResultsNode(resultsController: results, sectionNode: nil, objectNode: NCDatabaseCertificateRow.self))
	}
	
	//MARK: - TreeControllerDelegate
	
	override func treeController(_ treeController: TreeController, didSelectCellWithNode node: TreeNode) {
		super.treeController(treeController, didSelectCellWithNode: node)
		guard let row = node as? NCDatabaseCertificateRow else {return}
		Router.Database.CertificateInfo(certificate: row.object).perform(source: self, sender: treeController.cell(for: node))
	}
}
